[WIP][alpha] Create mirror QUIC listeners for HTTPS servers on gateways

su225 opened the pull request:
This PR demonstrates using QUIC in Istio. This is an experimental implementation where I have added a feature which creates mirror listener for an HTTPS listener at the gateway. **At the moment, it is not ideal and there are many issues and caveats**. I hope to iterate and refine it based on the feedback (relevant comment over the direction - https://github.com/istio/istio/issues/23018#issuecomment-755795067)

**Caveats, open questions, tasks to-do**
- [ ] Come up with a cleaner approach for this - maybe we can support QUIC protocol on the gateway for the users instead of doing shady things like this? API stuffs are not yet figured out
- [ ] Unit and integration testing these changes
- [ ] ALPN is not set properly. In envoy configs, it shows `h2` and `http 1.1` as it is just copied over from the filter chain config generated for TCP
- [ ] Refactor and clean up repetitive code (make it nicer)
- [ ] If the user wants both TCP+HTTP2 and UDP/QUIC+HTTP3 on the same port, then Kubernetes does not allow having two entries with same service->target port (even when protocols are different) for service of type `LoadBalancer`. (tested this in Kubernetes 1.21)
- [ ] If we have to create UDP listener to a different port specified by the user, then all things like authn, authz policies including ports should be translated as well. 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

su225 commented:
**Experiment**
Installed `IstioOperator` with the following settings on the latest main build of Istio (1.11-dev)
```yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: istio-with-quic
  namespace: istio-system
spec:
  meshConfig:
    accessLogFile: /dev/stdout
  hub: localhost:5000
  tag: istio-quic
  components:
    ingressGateways:
    - name: bookinfo-ingress
      namespace: bookinfo
      enabled: true
      label:
        gateway: bookinfo
        quic: enabled
      k8s:
        service:
          type: NodePort
          ports:
          - name: https-bookinfo
            port: 443
            targetPort: 8443
            protocol: TCP

          # Same service and target port with
          # the only difference being the protocol
          - name: http-quic-bookinfo
            port: 8443
            targetPort: 8443
            protocol: UDP
  values:
    global:
      imagePullPolicy: Always
      logging:
        level: "default:debug"
    pilot:
      env:
        PILOT_ENABLE_QUIC_LISTENERS: "true"
```

Then deployed the bookinfo application and setup routing (notice that there is no QUIC related stuffs in configs. QUIC mirror listener happens behind the scenes
```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: bookinfo-gw
  namespace: bookinfo
spec:
  selector:
    gateway: bookinfo
    quic: enabled
  servers:
  - port:
      number: 443
      protocol: HTTPS
      name: https-bookinfo
    hosts:
    - bookinfo.com
    tls:
      mode: SIMPLE
      credentialName: bookinfo-secret
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: bookinfo-vs
  namespace: bookinfo
spec:
  hosts:
  - bookinfo.com
  gateways:
  - bookinfo-gw
  http:
  - name: productpage-route
    route:
    - destination:
        host: productpage.bookinfo.svc.cluster.local
        port: { number: 9080 }
```

Finally test it with QUIC enabled cURL (it is not enabled by default. Instructions here - https://github.com/curl/curl/blob/master/docs/HTTP3.md). In my computer it is called `qcurl` which is built from main tree instead of release one. 
```
$ qcurl --resolve bookinfo.com:32409:172.18.0.2 -svk -H'Host: bookinfo.com' -o /dev/null --http3 https://bookinfo.com:32409
* Added bookinfo.com:32409:172.18.0.2 to DNS cache
* Hostname bookinfo.com was found in DNS cache
*   Trying 172.18.0.2:32409...
* Connect socket 5 over QUIC to 172.18.0.2:32409
* Sent QUIC client Initial, ALPN: h3,h3-29,h3-28,h3-27
* Connected to bookinfo.com () port 32409 (#0)
* h3 [:method: GET]
* h3 [:path: /]
* h3 [:scheme: https]
* h3 [:authority: bookinfo.com]
* h3 [user-agent: curl/7.78.0-DEV]
* h3 [accept: */*]
* Using HTTP/3 Stream ID: 0 (easy handle 0x17e12e0)
> GET / HTTP/3
> Host: bookinfo.com
> user-agent: curl/7.78.0-DEV
> accept: */*
> 
< HTTP/3 200
< content-type: text/html; charset=utf-8
< content-length: 1683
< server: istio-envoy
< date: Sat, 03 Jul 2021 15:41:18 GMT
< x-envoy-upstream-service-time: 7
< 
```
And Eureka!

Access logs from the gateway
```
[2021-07-03T15:36:26.271Z] "GET / HTTP/3" 200 - via_upstream - "-" 0 1683 22 21 "10.244.0.1" "curl/7.78.0-DEV" "99c3438c-1a1d-490d-b4ee-1551575c24a1" "bookinfo.com" "10.244.0.16:9080" outbound|9080||productpage.bookinfo.svc.cluster.local 10.244.0.29:55838 10.244.0.29:8443 10.244.0.1:29394 bookinfo.com productpage-route
[2021-07-03T15:36:37.365Z] "GET / HTTP/3" 200 - via_upstream - "-" 0 1683 9 7 "10.244.0.1" "curl/7.78.0-DEV" "410c1c4f-e5ca-4791-b55c-0a8be70c3509" "bookinfo.com" "10.244.0.16:9080" outbound|9080||productpage.bookinfo.svc.cluster.local 10.244.0.29:55840 10.244.0.29:8443 10.244.0.1:34725 bookinfo.com productpage-route
[2021-07-03T15:41:18.330Z] "GET / HTTP/3" 200 - via_upstream - "-" 0 1683 9 7 "10.244.0.1" "curl/7.78.0-DEV" "d89e444b-a33c-44d8-99ab-0e170e11a6f3" "bookinfo.com" "10.244.0.16:9080" outbound|9080||productpage.bookinfo.svc.cluster.local 10.244.0.29:55844 10.244.0.29:8443 10.244.0.1:57109 bookinfo.com productpage-route
```
Then I wanted to make sure that it is indeed sending UDP packets (it should). So tested with `tcpdump` (br-9dea9a86894f is the bridge to which my kind cluster is connected. So any packet entering the cluster is routed through this. Confirmed through `ip route get` command). Indeed there is some encrypted gibberish
```
$ sudo tcpdump -i br-9dea9a86894f udp -vv -X
21:11:18.318757 IP (tos 0x0, ttl 64, id 11170, offset 0, flags [DF], proto UDP (17), length 1228)
    su225.36820 > 172.18.0.2.32409: [bad udp cksum 0x5cf1 -> 0xe58c!] UDP, length 1200
	0x0000:  4500 04cc 2ba2 4000 4011 b257 ac12 0001  E...+.@.@..W....
	0x0010:  ac12 0002 8fd4 7e99 04b8 5cf1 cb00 0000  ......~...\.....
	0x0020:  0110 97a0 23b4 2892 c421 8857 01ca a61c  ....#.(..!.W....
	0x0030:  bf8a 1439 b96a a2fd eef7 f37f 6429 94d7  ...9.j......d)..
	0x0040:  5590 ac3e 3910 a700 4123 a926 b9ec 82fe  U..>9...A#.&....
	0x0050:  b0bd f4c1 945f 05ca 51c2 051a 1cd3 1d07  ....._..Q.......
```

Corresponding filter chain config in Envoy (it is long. So I have omitted many details. With the experiment you can inspect with `istioctl proxy-config listeners` command)
```yaml
- address:
    socketAddress:
      address: 0.0.0.0
      portValue: 8443
      protocol: UDP
  filterChains:
  - filterChainMatch:
      serverNames:
      - bookinfo.com
    filters:
    - name: envoy.filters.network.http_connection_manager
      typedConfig:
        '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
        codecType: HTTP3
        http3ProtocolOptions: {}
        httpFilters:
        - name: istio.metadata_exchange
        # Rest of HTTP filters go here
        - name: envoy.filters.http.router
          typedConfig:
            '@type': type.googleapis.com/envoy.extensions.filters.http.router.v3.Router

    transportSocket:
      name: envoy.transport_sockets.quic
      typedConfig:
        '@type': type.googleapis.com/envoy.extensions.transport_sockets.quic.v3.QuicDownstreamTransport
        downstreamTlsContext:
          commonTlsContext:
            alpnProtocols:
            - h2
            - http/1.1
            tlsCertificateSdsSecretConfigs:
            - name: kubernetes://bookinfo-secret
              sdsConfig:
                ads: {}
                resourceApiVersion: V3
          requireClientCertificate: false
  name: quic_0.0.0.0_8443
  reusePort: true
  trafficDirection: OUTBOUND
  udpListenerConfig:
    quicOptions: {}

```

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 8b8853e6daa0c5e67d029e689abac3ac2076244a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1411356092981579776) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 8b8853e6daa0c5e67d029e689abac3ac2076244a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1411356092981579776) | `/test release-notes_istio`
unit-tests_istio | 8b8853e6daa0c5e67d029e689abac3ac2076244a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/unit-tests_istio/1411356086312636416) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
Nice! I haven't had a chance to review this yet but wanted to point you to https://github.com/howardjohn/istio/tree/pilot/http3-gateway where I have done similar experiments

su225 commented:
Thanks @howardjohn - Your implementation is much simpler and understandable :) With some modification, it would be possible to support HTTP3 only ports, although the dominant use case is running both TCP+TLS and QUIC on the same port. Plus, I want to fix some stuffs like proper ALPN (h3 instead of h2 although from my yesterday's experiment it looked fine). 

However, some changes though (not major ones) - I wouldn't call them `UDPServer` although they are on UDP ports, but instead `QUICServer`. QUIC is NOT UDP although it uses it - one of the main reasons being not to scare middleboxes according to the original paper. It is a separate transport protocol than TCP and UDP (it looks closer than TCP in the sense that it provides some order at stream level, then flow and congestion control etc). I initially thought of including UDP, but that would be more complex and so reduced the scope to QUIC at gateways. (Hence the branch name). 

From [RFC 9000](https://datatracker.ietf.org/doc/html/rfc9000)
>  QUIC is a connection-oriented protocol that creates a stateful interaction between a client and server.

There seem to be some problems at Kubernetes layer though. Looks like it does not allow both TCP and UDP for the same port when service type is LoadBalancer.

su225 commented:
**UPDATE:** Turns out that Kubernetes has an alpha feature gate called `MixedProtocolLBService` to support the use cases mentioned. But then there is one more bug in Istio operator. So I opened - https://github.com/istio/istio/issues/33841 . Until then I would be using `NodePort` (not ideal. But fixing that bug + MixedProtocolLBService = ideal for QUIC)

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 8b8853e6daa0c5e67d029e689abac3ac2076244a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/unit-tests_istio/1411356086312636416) | `/test unit-tests_istio`
release-notes_istio | 55159e2155c26719e830052050be31db3999ba8a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1413916860872658944) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

su225 commented:
@howardjohn - I re-started with your code, fixed some issues and refactored the code. I also added some basic unit tests. I observed that Envoy is buggy for QUIC and not all settings worked. In fact, they mention that it is still alpha. So this definitely needs integration tests. Unfortunately, this is blocked by https://github.com/istio/istio/pull/33930 . For my experiments, I built `istioctl` from the branch containing the fix (#33930) and using it and here are the results

Notice that there are no NodePort hacks this time. It is the service of type `LoadBalancer` with TCP and UDP exposed on 443. I'm also thinking of adding `alt-svc` header for the response if QUIC is enabled on that port instead of user having to configure `VirtualService` for it.

For QUIC on port 443
```
$ qcurl -sv --cacert bookinfo.crt -o /dev/null --resolve bookinfo.com:443:172.18.210.1 --http3 https://bookinfo.com
* Added bookinfo.com:443:172.18.210.1 to DNS cache
* Hostname bookinfo.com was found in DNS cache
*   Trying 172.18.210.1:443...
* Connect socket 5 over QUIC to 172.18.210.1:443
* Sent QUIC client Initial, ALPN: h3,h3-29,h3-28,h3-27
* Connected to bookinfo.com () port 443 (#0)
* h3 [:method: GET]
* h3 [:path: /]
* h3 [:scheme: https]
* h3 [:authority: bookinfo.com]
* h3 [user-agent: curl/7.78.0-DEV]
* h3 [accept: */*]
* Using HTTP/3 Stream ID: 0 (easy handle 0x690290)
> GET / HTTP/3
> Host: bookinfo.com
> user-agent: curl/7.78.0-DEV
> accept: */*
> 
< HTTP/3 200
< content-type: text/html; charset=utf-8
< content-length: 1683
< server: istio-envoy
< date: Sat, 10 Jul 2021 17:48:03 GMT
< x-envoy-upstream-service-time: 8
```
For HTTP-over-TCP on port 443
```
* Added bookinfo.com:443:172.18.210.1 to DNS cache
* Hostname bookinfo.com was found in DNS cache
*   Trying 172.18.210.1:443...
* Connected to bookinfo.com (172.18.210.1) port 443 (#0)
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: bookinfo.crt
*  CApath: none
} [5 bytes data]
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.2 (IN), TLS handshake, Server hello (2):
...... Rest of TLS handshake omitted
.......
.......
< HTTP/1.1 200 OK
< content-type: text/html; charset=utf-8
< content-length: 1683
< server: istio-envoy
< date: Sat, 10 Jul 2021 17:49:24 GMT
< x-envoy-upstream-service-time: 3
< 
```

su225 commented:
Oh! I just noticed they have an issue already - https://github.com/envoyproxy/envoy/issues/15953

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | a351438e0e3885b43f69756a41a8ba83f4c5341e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1413924447282270208) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | c3ebff8601eb9b7dd01e4e223d7d263cabfa38b2 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1414143305779777536) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

su225 commented:
**UPDATE:** When HTTPS over TCP is used, it sends `alt-svc` header for the same port (currently hard-coded to the same port and constant max age of 1 day). The port is service port 443 (where pod is listening on 8443) as the client expects
```
[istio-quic]$ qcurl -o /dev/null -sv --cacert bookinfo-ca.crt --resolve bookinfo.com:443:172.18.210.1 https://bookinfo.com
#.... Lots of verbose stuffs
#
#
< HTTP/1.1 200 OK
< content-type: text/html; charset=utf-8
< content-length: 1683
< server: istio-envoy
< date: Wed, 14 Jul 2021 17:53:45 GMT
< x-envoy-upstream-service-time: 26
< alt-svc: h3=":443"; ma=86400; h3-29=":443"; ma=86400
```
Trying the same with H3/QUIC
```
[istio-quic]$ qcurl -o /dev/null -sv --cacert bookinfo-ca.crt --resolve bookinfo.com:443:172.18.210.1 --http3 https://bookinfo.com
* Added bookinfo.com:443:172.18.210.1 to DNS cache
* Hostname bookinfo.com was found in DNS cache
*   Trying 172.18.210.1:443...
* Connect socket 5 over QUIC to 172.18.210.1:443
* Sent QUIC client Initial, ALPN: h3,h3-29,h3-28,h3-27
* Connected to bookinfo.com () port 443 (#0)
* h3 [:method: GET]
* h3 [:path: /]
* h3 [:scheme: https]
* h3 [:authority: bookinfo.com]
* h3 [user-agent: curl/7.78.0-DEV]
* h3 [accept: */*]
* Using HTTP/3 Stream ID: 0 (easy handle 0x1cdc290)
> GET / HTTP/3
> Host: bookinfo.com
> user-agent: curl/7.78.0-DEV
> accept: */*
> 
< HTTP/3 200
< content-type: text/html; charset=utf-8
< content-length: 1683
< server: istio-envoy
< date: Wed, 14 Jul 2021 17:54:39 GMT
< x-envoy-upstream-service-time: 8
```

istio-testing commented:
@su225: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1417798682929008640) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1417798682929008640) | `/test release-notes_istio`
integ-security-k8s-tests_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-k8s-tests_istio/1417798676339757056) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1417798682929008640) | `/test release-notes_istio`
integ-security-k8s-tests_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-k8s-tests_istio/1417798676339757056) | `/test integ-security-k8s-tests_istio`
integ-security-multicluster-tests_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1417798678743093248) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-k8s-tests_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-k8s-tests_istio/1417798676339757056) | `/test integ-security-k8s-tests_istio`
integ-security-multicluster-tests_istio | 77f258028e85c538ec88c9d98cc663464b159d9b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1417798678743093248) | `/test integ-security-multicluster-tests_istio`
release-notes_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1418892907934388224) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1418892907934388224) | `/test release-notes_istio`
integ-security-multicluster-tests_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1418892903740084224) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1418892903740084224) | `/test integ-security-multicluster-tests_istio`
release-notes_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1420282507324035072) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1418892903740084224) | `/test integ-security-multicluster-tests_istio`
release-notes_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1420282507324035072) | `/test release-notes_istio`
lint_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/lint_istio/1420282505642119168) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1418892903740084224) | `/test integ-security-multicluster-tests_istio`
release-notes_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1420282507324035072) | `/test release-notes_istio`
lint_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/lint_istio/1420282505642119168) | `/test lint_istio`
unit-tests_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/unit-tests_istio/1420282500608954368) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 811bcd0747f9b2eabc1d431b68bc55cbf4f112d5 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1418892903740084224) | `/test integ-security-multicluster-tests_istio`
lint_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/lint_istio/1420282505642119168) | `/test lint_istio`
unit-tests_istio | fa9621839890c806f7e5b84c1bc9c3a0e8c8408b | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/unit-tests_istio/1420282500608954368) | `/test unit-tests_istio`
release-notes_istio | 1194963ba9d4c817db12ee137a7d0b67e16a16f0 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/release-notes_istio/1420295392397889536) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 36b0a5807231ec512dc7e118568f843ea73dc10e | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1420297046354563072) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -120,96 +119,157 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
+						server.GetPort().GetNumber(), server.GetPort().GetName(), server.GetPort().GetProtocol())
+					routeName := mergedGateway.TLSServerInfo[server].RouteName
+					if mergedGateway.TLSToQUICServerRouteMap[routeName] != "" {
+						routeName = mergedGateway.TLSToQUICServerRouteMap[routeName]
```
is `mergedGateway.TLSToQUICServerRouteMap[routeName]` ever `""` here?

su225 replied:
I originally wrote something and this remained. I changed the name and type to set. It is the set of routes on which H3 has to be advertised (by setting `alt-svc` header)

howardjohn commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -120,96 +119,157 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
+						// Not doing so is a security hole as would allow bypassing auth.
+						ListenerProtocol:   istionetworking.ListenerProtocolHTTP,
+						TransportProtocol:  istionetworking.TransportProtocolQUIC,
+						IstioMutualGateway: false, // Currently, we don't support this. Revisit later
```
just curious, why not?

su225 replied:
Simple answer - I didn't know much about that. I vaguely knew it is about mTLS. TBH searching for its usages, I still don't understand why it is there. So I set it to `false` and moved on so that I can focus on getting simple TLS to work.

howardjohn commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -229,6 +289,16 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
+	case istionetworking.TransportProtocolTCP:
+		return bind + "_" + strconv.Itoa(port)
+	case istionetworking.TransportProtocolQUIC:
+		return "quic_" + bind + "_" + strconv.Itoa(port)
```
we may consider here, and many other places replacing quic with udp. The rationale is that in the future we may add non-quic UDP and it would be on the same listener?

Unless quic + non-quic UDP on the same listener will never be supported by envoy

su225 replied:
If I understand it correctly [RFC-9000](https://datatracker.ietf.org/doc/html/rfc9000#section-17.3.1) gives a hint on QUIC co-existing with other UDP based protocols on the same port. 

>    Fixed Bit:  The next bit (0x40) of byte 0 is set to 1.  Packets
      containing a zero value for this bit are not valid packets in this
      version and MUST be discarded.  A value of 1 for this bit allows
      QUIC to coexist with other protocols; see [RFC7983].

So I agree with you that although QUIC is connection-oriented protocol, it is still UDP now. So I changed it to `udp_` as you suggested. I know that these names show up on `istioctl proxy-config listeners`. Question - Is this something that needs to be backwards-compatible?

howardjohn replied:
> Question - Is this something that needs to be backwards-compatible?

No, but if you change it it would probably make upgrades not work because Envoy would complain it has two listeners on the same port. But aside from that users are not supposed to depend on XDS details

su225 replied:
Ok. I thought we send the entire state of the world on each LDS push - meaning, if the new update received does not have the name then envoy would clean up the listener. Hmm...That does not seem to be the case here? (I don't know the details of Envoy XDS implementation). Anyways, I changed it to `udp_` as suggested.

howardjohn replied:
We do but envoy doesn't actually atomically apply them like that IIUC. So the new ones still can conflict with old ones

su225 replied:
Oh! Ok. Learned something new :) I didn't know "does not atomically apply them" part. I guess even if it applies atomically, Envoy has to put old listeners to draining mode for deletion (socket still not closed at that time). For modification, simple way is delete+add - in this case we will run into port-conflict.

But still - while upgrading don't we restart pods for newer version of sidecar to be injected and it connects to newer revision of istiod. Due to restart, envoy should have forgotten configs and starts with clean slate. Right? I see that the problem occurs when someone upgrades it in-place (which I think is not recommended) and sidecar reconnects. Another possibility is somebody changing Istiod tag (due to say, hot fix) resulting in newer Istiod sending something (again - hot-fix should be avoided).

Anyways, I have changed it to `udp_` as you suggested already. Still not fully convinced with "upgrades does not work" part. I have outlined scenarios that I could think of and yes - I agree that it would be a problem in case of in-case upgrades as for the sidecar it would be "same istiod address and reconnect" resulting in mixing up of envoy config state. But for canary upgrade scenario it should not cause issues as the pod would be restarted to point to the new revision.

howardjohn commented on pilot/pkg/model/gateway.go:
```diff
@@ -247,23 +260,58 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy) *MergedGateway
+						if features.EnableQUICListeners && gateway.IsEligibleForHTTP3Upgrade(s) &&
+							udpSupportedPort(s.GetPort().GetNumber(), gwAndInstance.instances) {
+							log.Debugf("Server at port %d eligible for HTTP3 upgrade. Add QUIC listener", serverPort.Number)
+							h3MirrorRouteName := "h3-mirror." + routeName
```
Oh for alt-svc... 2x routes is fairly expensive. We may consider adding support directly to envoy (in HCM) to add this - I think there was high level support for that

su225 replied:
If we would like to reduce the number of routes, how about `alt-svc` in both cases? I get it - it is not needed when the client talks H3 and it is redundant. According to the RFC-7838 that is something the client can ignore.

From https://datatracker.ietf.org/doc/html/rfc7838#section-2.4
>    By their nature, alternative services are OPTIONAL: clients do not need to use them.

> Therefore, if a client supporting this specification becomes aware of an alternative service, the client SHOULD use that alternative service for all requests to the associated origin as soon as it is available

su225 replied:
Just noticed that cloudflare sends `alt-svc` even for H3. So I assume this shouldn't be a problem in general!

howardjohn commented on tests/integration/security/sds_ingress/ingress_test.go:
```diff
@@ -223,15 +223,15 @@ func TestSingleMTLSGateway_CompoundSecretRotation(t *testing.T) {
 						Cert:       ingressutil.TLSClientCertA,
 					}
-					ingressutil.SendRequestOrFail(t, ing, host, credName[0], ingressutil.Mtls, tlsContext,
+					ingressutil.SendRequestOrFail(t, ing, host, credName[0], ingressutil.Mtls, tlsContext, false, /* QUIC */
```
would prefer to make a special SendQUICRequestOrFail or something, to avoid changing all of these.

su225 replied:
Done.

howardjohn commented on prow/config/trustworthy-jwt.yaml:
```diff
@@ -1,6 +1,8 @@
 kind: Cluster
 apiVersion: kind.x-k8s.io/v1alpha4
+featureGates:
+  MixedProtocolLBService: true
```
I think we need to be careful to ensure we don't try to run these tests on old Kubernetes versions (which run in postusbmit).

will unknown featureGates be silently ignored or fail hard? If they are silent, we can just use RequireMaxVersion in the test. if fail hard, need some more logic in test script I think to pick the config

howardjohn replied:
Oh you already did the RequireMaxVersion part

su225 replied:
Damn! I realised I had issed `!` which resulted in tests skipping when version >= 1.21. Fixed it :D

su225 replied:
We have a problem. For kind version 1.19 (possibly below that), the feature gate causes issues. So post-submits blow up. Gotta do something on that. :(

su225 replied:
Ok. I need to dig into `test-infra`. 🤦🏼

su225 replied:
I have fixed the test infra by turning on `MixedProtocolLBSupport` for K8s >= 1.20 as suggested. Although it is not yet merged (hoping it would be merged), I'll mark this resolved.

su225 commented on pilot/pkg/model/gateway.go:
```diff
@@ -56,8 +56,19 @@ type MergedGateway struct {
-	MergedServers map[ServerPort]*MergedServers
+	// MergedTCPServers map from physical port to virtual servers
+	// using TCP protocols (like HTTP1.1, H2, mysql, redis etc)
+	MergedTCPServers map[ServerPort]*MergedServers
```
Ok. I changed it to `MergedServers` which is what was there before and `MergedQUICTransportServers`.

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-security-multicluster-tests_istio | 29d3ad71e50d76620657b7025577efa139301321 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1421175602953588736) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

su225 commented:
/test integ-security-multicluster-tests_istio

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | c620d1f5d1fad02eece69b812cdfa3f1d8b16cb8 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/lint_istio/1422950109145993216) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | cb1568284ad95598c95732b29b7bb4d30f219112 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-pilot-k8s-tests_istio/1423317045335822336) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | cb1568284ad95598c95732b29b7bb4d30f219112 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-pilot-k8s-tests_istio/1423317045335822336) | `/test integ-pilot-k8s-tests_istio`
integ-security-k8s-tests_istio | cb1568284ad95598c95732b29b7bb4d30f219112 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-k8s-tests_istio/1423317045423902720) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@su225: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-pilot-k8s-tests_istio | cb1568284ad95598c95732b29b7bb4d30f219112 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-pilot-k8s-tests_istio/1423317045335822336) | `/test integ-pilot-k8s-tests_istio`
integ-security-k8s-tests_istio | cb1568284ad95598c95732b29b7bb4d30f219112 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-k8s-tests_istio/1423317045423902720) | `/test integ-security-k8s-tests_istio`
integ-security-multicluster-tests_istio | cb1568284ad95598c95732b29b7bb4d30f219112 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/integ-security-multicluster-tests_istio/1423317047776907264) | `/test integ-security-multicluster-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

su225 commented:
Linking fortio issue for cross-reference - https://github.com/fortio/fortio/issues/455

howardjohn commented:
> But still - while upgrading don't we restart pods for newer version of
sidecar to be injected and it connects to newer revision of istiod. Due to
restart, envoy should have forgotten configs and starts with clean slate.
Right? I see that the problem occurs when someone upgrades it in-place
(which I think is not recommended) and sidecar reconnects. Another
possibility is somebody changing Istiod tag (due to say, hot fix) resulting
in newer Istiod sending something (again - hot-fix should be avoided).

Yes it only matters if Istiod is hot-upgraded. Not recommended but still
done by a fairly large number of users unfortunately.

BTW - even if they do hot-upgrade Istiod, there is still a way to make it
work - we can conditionally send certain XDS configs based on proxy
version, so it's deterministic.

Anyhow probably not a big deal in this case

On Sat, Aug 7, 2021 at 9:17 AM Suchith J N ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In pilot/pkg/networking/core/v1alpha3/gateway.go
> <https://github.com/istio/istio/pull/33817#discussion_r684650649>:
>
> > @@ -229,6 +289,16 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
>  	return builder
>  }
>
> +func (configgen *ConfigGeneratorImpl) getListenerName(bind string, port int, transport istionetworking.TransportProtocol) string {
> +	switch transport {
> +	case istionetworking.TransportProtocolTCP:
> +		return bind + "_" + strconv.Itoa(port)
> +	case istionetworking.TransportProtocolQUIC:
> +		return "quic_" + bind + "_" + strconv.Itoa(port)
>
> Oh! Ok. Learned something new :) I didn't know "does not atomically apply
> them" part. I guess even if it applies atomically, Envoy has to put old
> listeners to draining mode for deletion (socket still not closed at that
> time). For modification, simple way is delete+add - in this case we will
> run into port-conflict.
>
> But still - while upgrading don't we restart pods for newer version of
> sidecar to be injected and it connects to newer revision of istiod. Due to
> restart, envoy should have forgotten configs and starts with clean slate.
> Right? I see that the problem occurs when someone upgrades it in-place
> (which I think is not recommended) and sidecar reconnects. Another
> possibility is somebody changing Istiod tag (due to say, hot fix) resulting
> in newer Istiod sending something (again - hot-fix should be avoided).
>
> Anyways, I have changed it to udp_ as you suggested already. Still not
> fully convinced with "upgrades does not work" part. I have outlined
> scenarios that I could think of and yes - I agree that it would be a
> problem in case of in-case upgrades as for the sidecar it would be "same
> istiod address and reconnect" resulting in mixing up of envoy config state.
> But for canary upgrade scenario it should not cause issues as the pod would
> be restarted to point to the new revision.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/33817#discussion_r684650649>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXI43E65CZTT6SQYJODT3VMBFANCNFSM47YISZ5Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

hzxuzhonghu commented:
Haven't tried quic yet.

istio-testing commented:
@su225: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

hzxuzhonghu commented:
make sense, it is safe AFAIK

su225 commented:
> Is it safe when the client does not support http3? Is it a must the gateway will require http3 for the second request?

@hzxuzhonghu - Yes. It is safe. This creates a QUIC mirror listener for every HTTPS port on the gateway provided there is a UDP port opened on the gateway with the same number. (Like 443/TCP and 443/UDP). Existing listeners are not removed. With `Alt-Svc`, we tell the client that they can use HTTP/3, but it is not mandatory, but additional choices. They can continue to use HTTP/1.1 or HTTP/2 over TCP.

su225 commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -120,96 +119,157 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
+			istionetworking.TransportProtocolQUIC: mergedGateway.MergedQUICTransportServers,
+		}
+
+		for transport, gwServers := range transportToServers {
```
Much of the code - like building listener options, calling plugins to create Auth filters, code to generate listeners and filter chains are common to both. The difference is in building HTTP related filter chains. I have moved TCP and QUIC sections of the switch case to its own functions for clarity.

su225 commented on pilot/pkg/model/gateway.go:
```diff
@@ -273,16 +315,30 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy) *MergedGateway
 }
 
+func udpSupportedPort(number uint32, instances []*ServiceInstance) bool {
+	log.Debugf("udpSupportedPort:  -> checking for port %d", number)
```
Done.

su225 commented on pilot/pkg/model/gateway.go:
```diff
@@ -273,16 +315,30 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy) *MergedGateway
+func udpSupportedPort(number uint32, instances []*ServiceInstance) bool {
+	log.Debugf("udpSupportedPort:  -> checking for port %d", number)
+	for _, w := range instances {
+		log.Debugf("udpSupportedPort: ---> checking with instance: port=%d, protocol=%s",
```
Done.

su225 commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -1417,27 +1440,50 @@ func buildListener(opts buildListenerOpts, trafficDirection core.TrafficDirectio
+		// TODO: switch on TransportProtocolQUIC is in too many places now. Once this is a bit
+		//       mature, refactor some of these to an interface so that they kick off the process
+		//       of building listener, filter chains, serializing etc based on transport protocol
+		listenerName := "udp_" + opts.bind + "_" + strconv.Itoa(opts.port.Port)
```
Done.

su225 commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -536,9 +540,34 @@ func translateRoute(node *model.Proxy, in *networking.HTTPRoute,
 }
 
+func buildH3AltSvcHeader(port int, h3Alpns []string) *core.HeaderValueOption {
+	headerName := "Alt-svc"
```
Changed it to `alt-svc`.

su225 commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -536,9 +540,34 @@ func translateRoute(node *model.Proxy, in *networking.HTTPRoute,
 	}
 
+	if isH3DiscoveryNeeded {
+		h3DiscoveryHeader := buildH3AltSvcHeader(port, util.ALPNHttp3OverQUIC)
```
I'm not sure if Envoy does that. I have to check.

hzxuzhonghu replied:
I am not sure, just a question here

howardjohn replied:
suggest adding note to https://github.com/envoyproxy/envoy/issues/15953

su225 replied:
I added a note about our usage.

su225 commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -536,9 +540,34 @@ func translateRoute(node *model.Proxy, in *networking.HTTPRoute,
+		// Max-age is hardcoded to 1 day for now.
+		valParts = append(valParts, fmt.Sprintf(`%s=":%d"; ma=86400`, alpn, port))
+	}
+	headerVal := strings.Join(valParts, "; ")
```
Fixed it. Also added an example comment on what is returned for `www.cloudflare.com`
```
alt-svc: h3-27=":443"; ma=86400, h3-28=":443"; ma=86400, h3-29=":443"; ma=86400, h3=":443"; ma=86400
```

hzxuzhonghu commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -289,7 +220,94 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 }
 
-func (configgen *ConfigGeneratorImpl) getListenerName(bind string, port int, transport istionetworking.TransportProtocol) string {
+func (configgen *ConfigGeneratorImpl) getTCPTransportBasedFilterChains(
```
prefer buildGatewayTCPBasedFilterChains

hzxuzhonghu commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -289,7 +220,94 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
+	return newFilterChains
+}
+
+func (configgen *ConfigGeneratorImpl) getHTTP3FilterChains(
```
buildGatewayHTTP3FilterChains

su225 replied:
Done.

hzxuzhonghu commented:
lgtm, @howardjohn @ramaraochavali to put more eyes on it.

howardjohn commented on pilot/pkg/model/gateway.go:
```diff
@@ -253,6 +265,23 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy) *MergedGateway
+						// We have TLS settings defined and we have already taken care of unique route names
+						// if it is HTTPS. So we can construct a QUIC server on the same port. It is okay as
+						// QUIC listens on UDP port, not TCP
+						if features.EnableQUICListeners && gateway.IsEligibleForHTTP3Upgrade(s) &&
```
nit: IsEligibleForHTTP3Upgrade already checks EnableQUICListeners, no need to replace

su225 replied:
Yeah. I see it is redundant. Guess it is fine for readability?

howardjohn commented on pilot/pkg/features/pilot.go:
```diff
@@ -529,6 +529,9 @@ var (
 	).Get()
+
+	EnableQUICListeners = env.RegisterBoolVar("PILOT_ENABLE_QUIC_LISTENERS", false,
+		"If true, QUIC listeners will be generated wherever there are listeners terminating TLS on gateways").Get()
```
This is missing `... if the gateway Service exposes a UDP port`

su225 replied:
Done. Added.

howardjohn commented on pilot/pkg/model/gateway.go:
```diff
@@ -253,6 +265,23 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy) *MergedGateway
+							udpSupportedPort(s.GetPort().GetNumber(), gwAndInstance.instances) {
+							log.Debugf("Server at port %d eligible for HTTP3 upgrade. Add UDP listener for QUIC", serverPort.Number)
+							if mergedQUICServers[serverPort] == nil {
+								mergedQUICServers[serverPort] = &MergedServers{
```
very nit: prefer setting it to empty MergedServers then do not have an `else` - always set it. That way its consistent code if we need a change

su225 replied:
Done

howardjohn commented on pilot/pkg/model/gateway.go:
```diff
@@ -253,6 +265,23 @@ func MergeGateways(gateways []gatewayWithInstances, proxy *Proxy) *MergedGateway
+							if mergedQUICServers[serverPort] == nil {
+								mergedQUICServers[serverPort] = &MergedServers{
+									Servers:   []*networking.Server{s},
+									RouteName: routeName,
```
what if routeName changes between serverPort? Is it possible?

su225 replied:
Something I need to go through once again.

su225 replied:
It is possible because different HTTPS servers would have different route names. I went through the code again and found that `RouteName` is unnecessary for HTTPS and HTTP/3 cases. This is because we take it from `TLSServerInfo`. The field is used only for plaintext HTTP. Hence I removed this in both places.

howardjohn commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -1184,6 +1185,8 @@ type httpListenerOpts struct {
 	addGRPCWebFilter bool
 	useRemoteAddress bool
+
+	supportHTTP3 bool
```
I think this needs a comment/rename. "Support" seems to imply it _additionally_ supports HTTP3 - not that it is _only_ HTTP3

su225 replied:
I changed it to `http3Only`.

howardjohn commented on pilot/pkg/networking/core/v1alpha3/listener.go:
```diff
@@ -1320,18 +1329,23 @@ func buildHTTPConnectionManager(listenerOpts buildListenerOpts, httpOpts *httpLi
 	var listenerFilters []*listener.ListenerFilter
 
 	// add a TLS inspector if we need to detect ServerName or ALPN
+	// (this is not applicable for QUIC listeners)
```
Because its always inspecting the TLS?

su225 replied:
Ideally, when UDP and QUIC are multiplexed on the same port, then a listener filter like `QuicInspector` (counterpart of `TlsInspector`) would be needed to decide whether to create `QuicTransportSocket` or `UdpTransportSocket`. However, we are not supporting UDP in Istio at the moment. So it is fine not to have it. That said, I'm not sure if Envoy's `TlsInspector` works with UDP sockets (I guess it does not support DTLS)

howardjohn commented on pilot/pkg/networking/core/v1alpha3/route/route.go:
```diff
@@ -536,9 +540,36 @@ func translateRoute(node *model.Proxy, in *networking.HTTPRoute,
 	return out
 }
 
+func buildH3AltSvcHeader(port int, h3Alpns []string) *core.HeaderValueOption {
```
Add a unit test for this

su225 replied:
Done.

howardjohn commented on pilot/pkg/networking/util/util.go:
```diff
@@ -121,6 +126,9 @@ var ALPNInMeshWithMxc = []string{"istio-peer-exchange", "istio"}
 var ALPNHttp = []string{"h2", "http/1.1"}
 
+// ALPNHttp3OverQUIC advertises that Proxy is going to talk HTTP/3 over QUIC
+var ALPNHttp3OverQUIC = []string{"h3", "h3-29"} // TODO: Add various h3-* corresponding to drafts
```
Do we need anything other than `h3`? Seems like the old ones are just legacy, since we are greenfield deployment we can just support the official RFC?

howardjohn replied:
Easier to add than to take away :slightly_smiling_face:

su225 replied:
Yeah. I'll remove `h3-29`

howardjohn commented on tests/integration/security/sds_ingress/util/util.go:
```diff
@@ -289,8 +290,61 @@ func SendRequestOrFail(ctx framework.TestContext, ing ingress.Instance, host str
+	ing.CallWithRetryOrFail(ctx, opts, retry.Timeout(time.Minute*2))
+}
+
+func SendQUICRequestOrFail(ctx framework.TestContext, ing ingress.Instance, host string, path string,
```
seems we can reduce duplication with SendRequestOrFail a lot

su225 replied:
I guess it was something similar originally. `SendRequestOrFail` is used in many places. However, I'll still change it even though the diff would be more - it would be trivial anyways.

howardjohn replied:
What I would do is:

```
SendRequestOrFail () { internalSend(protocol=TCP) }
SendQUICRequestOrFail() {internalSend(protocol=QUIC) }
internalSend() { .. existing logic ...}
```

su225 replied:
Ah! Right! This is neat.

howardjohn commented on tests/integration/security/sds_ingress_k8sca/quic/ingress_test.go:
```diff
@@ -0,0 +1,98 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package quic
```
IMO we don't need this one. `k8sca` is a variant on `sds_ingress`. We don't need all combinatoric explosion of possible options - `k8sca` and `quic` have no relation and are extremely unlikely to conflict

su225 replied:
Removed this. Looking at it, I guess only the PKI stuff changes. Correct? As long as we have the right CA certs for verification, it does not affect QUIC-TLS handshake.

howardjohn reviewed (commented):
LGTM! just a few nits

istio-testing commented:
@su225: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 0a6a2865d5fd37258fae67017204349eedbf33cd | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/33817/lint_istio/1432959830980562944) | `/test lint_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn reviewed (approved):
Awesome job! I am excited to see this land and impressed with the work done here.
